src/gtimelog/tests/core/test_reports.py

```diff
@@ -360,5 +361,71 @@ def doctest_reports_custom_range_report_categorized():
     """
 
 
+class TestReport(unittest.TestCase):
+    TEST_TIMELOG = textwrap.dedent("""
+        2014-05-27 10:03: arrived
+        2014-05-27 10:13: edx: introduce topic to new sysadmins -- edx
+        2014-05-27 10:30: email
+        2014-05-27 12:11: meeting: how to support new courses?  -- edx meeting
+        2014-05-27 15:12: edx: write test procedure for EdX instances -- edx sysadmin
+        2014-05-27 17:03: cluster: set-up accounts, etc. -- sysadmin hpc
+        2014-05-27 17:14: support: how to run statistics on Hydra? -- support hydra
+        2014-05-27 17:36: off: pause **
+        2014-05-27 17:38: email
+        2014-05-27 19:06: off: dinner & family **
+        2014-05-27 22:19: cluster: fix shmmax-shmall issue -- sysadmin hpc
+        """)
+
+    def setUp(self):
+        self.tw = make_time_window(
+            StringIO(self.TEST_TIMELOG),
+            datetime(2014, 5, 27, 9, 0),
+            datetime(2014, 5, 27, 23, 59),
+            time(2, 0),
+        )
+
+    def test_report_tags(self):
+        rp = Reports(self.tw)
+        txt = StringIO()
+        # use same tags as in tests above, so we know the totals
+        rp._report_tags(txt, ['meeting', 'hpc'])
+        self.assertEqual(
+            txt.getvalue().strip(),
+            textwrap.dedent("""
+            Time spent in each area:
+              hpc          5:04
+              meeting      1:41
+            Note that area totals may not add up to the period totals,
+            as each entry may be belong to multiple areas (or none at all).
+            """).strip())
+
+    def test_daily_report_includes_tags(self):
+        rp = Reports(self.tw)
+        txt = StringIO()
+        rp.daily_report(txt, 'me@example.com', 'me')
+        self.assertIn('Time spent in each area', txt.getvalue())
+
+    def test_weekly_report_includes_tags(self):
+        rp = Reports(self.tw)
+        txt = StringIO()
+        rp.weekly_report(txt, 'me@example.com', 'me')
+        self.assertIn('Time spent in each area', txt.getvalue())
+
+    def test_monthly_report_includes_tags(self):
+        rp = Reports(self.tw)
+        txt = StringIO()
+        rp.monthly_report(txt, 'me@example.com', 'me')
+        self.assertIn('Time spent in each area', txt.getvalue())
+
+    def test_categorized_report_includes_tags(self):
+        rp = Reports(self.tw, style='categorized')
+        txt = StringIO()
+        rp.weekly_report(txt, 'me@example.com', 'me')
+        self.assertIn('Time spent in each area', txt.getvalue())
+        txt = StringIO()
+        rp.monthly_report(txt, 'me@example.com', 'me')
+        self.assertIn('Time spent in each area', txt.getvalue())
+
+
 def test_suite():
-    return unittest.defaultTestLoader.loadTestsFromName(__name__)
+    return doctest.DocTestSuite(__name__, optionflags=doctest.NORMALIZE_WHITESPACE)
```

Comment:
This is incorrect: you need both, combined into one `unittest.TestSuite([...loadTestsFromName..., doctest.DocTestSuite(...)])`.

Or we could hasten the switch to pytest.   Ideally separately, in master, without the refactorings.

Reply:
This method works locally. As i'm not a expert of doctest, i think i don't understand what you mean. Migrating to pytest once refactoring is done could be a great idea.

Reply:
It does not work, and here's proof: apply this patch
```diff
index 5171dd3..cf38fe5 100644
--- a/src/gtimelog/tests/core/test_reports.py
+++ b/src/gtimelog/tests/core/test_reports.py
@@ -16,6 +16,8 @@ def doctest_reports_weekly_report_categorized():
         >>> minimum = datetime(2010, 1, 25)
         >>> maximum = datetime(2010, 1, 31)
 
+        >>> assert not 'doctest'
+
         >>> window = make_time_window(minimum=minimum, maximum=maximum)
         >>> reports = Reports(window)
         >>> reports.weekly_report_categorized(sys.stdout, 'foo@bar.com',
@@ -426,6 +428,8 @@ class TestReport(unittest.TestCase):
         rp.monthly_report(txt, 'me@example.com', 'me')
         self.assertIn('Time spent in each area', txt.getvalue())
 
+        assert not 'unittest'
+
 
 def test_suite():
     return doctest.DocTestSuite(__name__, optionflags=doctest.NORMALIZE_WHITESPACE)
```

and run `tox -e py38`.  You should see two failures, but you'll see only one.

Reply:
Just a simple question, do you want to merge this PR really ? As you want to perfect it. Is it not preferable to take this PR as a template and do PR's individually for the sake of our community ? (just see @andrewshadura critics about it)

Thanks for your return but it does not help more to well understand your patch. As locally all 95 tests works without errors :S Can you explain me further this patch please ? I need to learn sometimes ;)

Reply:
Yes, individual PRs would be preferable to one large PR doing everything.

As for this specific comment, the problem I'm trying to get at is that not all tests are being executed!  You said "all 95 tests pass", but there are actually 100 tests, which you'll see if you include both unit tests and doctests in the test suite.  The diff with 'assert' statements was meant to illustrate that some of the tests are not being run at all.

Reply:
Indeed, if I locally apply

```diff
diff --git a/src/gtimelog/tests/core/test_reports.py b/src/gtimelog/tests/core/test_reports.py
index 5171dd3..ca006ca 100644
--- a/src/gtimelog/tests/core/test_reports.py
+++ b/src/gtimelog/tests/core/test_reports.py
@@ -428,4 +428,7 @@ class TestReport(unittest.TestCase):
 
 
 def test_suite():
-    return doctest.DocTestSuite(__name__, optionflags=doctest.NORMALIZE_WHITESPACE)
+    return unittest.TestSuite([
+        doctest.DocTestSuite(optionflags=doctest.NORMALIZE_WHITESPACE),
+        unittest.defaultTestLoader.loadTestsFromName(__name__),
+    ])
```

then I see 100 tests running and one of them fails, due to missing blank lines.